gateway/metrics/exporter.go

```diff
@@ -72,9 +80,46 @@ func (e *Exporter) StartServiceWatcher(endpointURL url.URL, metricsOptions Metri
 	ticker := time.NewTicker(interval)
 	quit := make(chan struct{})
 
-	timeout := 3 * time.Second
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+
+				namespaces, err := e.getNamespaces(endpointURL)
+				if err != nil {
+					log.Println(err)
+				}
+
+				if len(namespaces) == 0 {
+					emptyNamespace := ""
+					services, err := e.getFunctions(endpointURL, emptyNamespace)
+					if err != nil {
+						log.Println(err)
+						continue
+					}
+					e.services = services
+				} else {
+					for _, namespace := range namespaces {
+						services, err := e.getFunctions(endpointURL, namespace)
```

Comment:
I think this works well for now, but I also suspect a larger discussion is needed about how/if we move this into the provider sending a request per namespace every second seems like something the provider should encapsulate so that it can optimize that behavior as it best makes sense for itself.  

This current implementation would create a ["thundering herd"](https://en.wikipedia.org/wiki/Thundering_herd_problem)-like problem with no way to really streamline or optimize it.   For example, the k8s provider should be able to request the existing functions from several/all namespaces in a single request.  

Perhaps we need a new endpoint in the provider that allows the gateway to send 0, 1, or more namespaces as part of the request. When it is 0, it would request for all namespaces, otherwise the response would send back a list of functions for the provided namespaces. Alternatively, we could allow a special `_all` value. Then 0 namespaces == default, 1 or more applied as obvious, `_all` return functions from all possible namespaces.

Reply:
I completely agree and we should move function aggregation logic to the provider. 

I think we can modify the exiting function list endpoint using some special query parameters  like `_all`, for which will combine functions from all namespaces and return it. Otherwise it works as it is working now. I think it will be a very minimal change compared to having a new endpoint. For new endpoint we will have to write different logic depending on the provider at several places.

Reply:
Introducing a thundering herd issue concerns me. What are the alternatives to doing that?

Reply:
Thundering herd issue is because every 5 second (configurable) it tries to get all namespaces and then for each namespace it lists its functions to count service/replica count of functions. 

Solution is to use a new endpoint or existing which list all functions across all namespaces so that we don't have to make these multiple calls. This implementation can be done in faas-netes which will be efficient.

Reply:
@viveksyngh can you write a bash loop to deploy 100-200 functions and then measure the effect of the patch?

Reply:
```bash
for i in {1..100} do ;
   faas-cli deploy --image figlet --name fn$i
done
```